Changes in module 'nf-core/modules/samtools/merge'
--- modules/nf-core/modules/samtools/merge/main.nf
+++ modules/nf-core/modules/samtools/merge/main.nf
@@ -25,7 +25,7 @@
     script:
     def args = task.ext.args   ?: ''
     prefix   = task.ext.prefix ?: "${meta.id}"
-    def file_type = input_files[0].getExtension()
+    def file_type = input_files[0].getExtension() ?: "bam"
     def reference = fasta ? "--reference ${fasta}" : ""
     """
     samtools \\

************************************************************
